File: android/app/src/main/java/com/jvoice/core/data/NewsCodec.kt
package com.jvoice.core.data

import com.google.firebase.firestore.FieldValue
import com.jvoice.core.i18n.LocalizedText
import com.jvoice.news.data.model.Category
import com.jvoice.news.data.model.NewsArticle
import com.jvoice.news.data.model.NewsStatus
import com.jvoice.news.data.model.NotificationItem
import com.jvoice.news.data.model.NotificationType
import com.jvoice.news.data.model.UserRole

/**
 * Firestore ↔ model for the News module.
 *
 * Enums are stored by `name` (`"PUBLISHED"`, not an ordinal). Ordinals are
 * smaller and completely unmaintainable: inserting a value into the middle of
 * [NewsStatus] would silently reinterpret every stored document.
 *
 * Timestamps are plain epoch millis rather than Firestore `Timestamp` objects,
 * because the models already use `Long` and the app's relative-time formatting
 * works on millis. The one exception is a *server* timestamp on create, which is
 * written as [FieldValue.serverTimestamp] and then read back as millis — see
 * [articleToMap].
 */

/* ================================================================== article */

fun NewsArticle.toMap(): Map<String, Any?> = mapOf(
    "headline" to headline.toMap(),
    "shortDescription" to shortDescription.toMap(),
    "content" to content.toMap(),
    "categoryId" to categoryId,
    "location" to location,
    "imageUrl" to imageUrl,
    "photoUrls" to photoUrls,
    "videoUrls" to videoUrls,
    "tags" to tags.toMapList(),
    "isBreaking" to isBreaking,
    "isFeatured" to isFeatured,
    "isTrending" to isTrending,
    "status" to status.name,
    "reporterId" to reporterId,
    "reporterName" to reporterName,
    "reporterAvatarUrl" to reporterAvatarUrl,
    "createdAt" to createdAt,
    "updatedAt" to updatedAt,
    "publishedAt" to publishedAt,
    "rejectionReason" to rejectionReason?.toMap(),
    "editorNote" to editorNote?.toMap(),
    "views" to views,
    "reportCount" to reportCount,
    // Denormalised so the reader feed can order published stories without
    // reading every draft: Firestore cannot order on a field some documents
    // lack, and publishedAt is null until publication.
    "sortAt" to (publishedAt ?: createdAt)
)

fun articleFrom(id: String, data: Map<String, Any?>): NewsArticle {
    val createdAt = data.long("createdAt")
    return NewsArticle(
        id = id,
        headline = localizedFrom(data["headline"]),
        shortDescription = localizedFrom(data["shortDescription"]),
        content = localizedFrom(data["content"]),
        categoryId = data.str("categoryId"),
        location = data.str("location"),
        imageUrl = data.str("imageUrl"),
        photoUrls = data.strList("photoUrls"),
        videoUrls = data.strList("videoUrls"),
        tags = localizedListFrom(data["tags"]),
        isBreaking = data.bool("isBreaking"),
        isFeatured = data.bool("isFeatured"),
        isTrending = data.bool("isTrending"),
        status = data.enum("status", NewsStatus.DRAFT),
        reporterId = data.str("reporterId"),
        reporterName = data.str("reporterName"),
        reporterAvatarUrl = data.str("reporterAvatarUrl"),
        createdAt = createdAt,
        // Defaults to createdAt rather than 0, so a record written without it
        // does not sort to the beginning of time.
        updatedAt = data.long("updatedAt", createdAt),
        publishedAt = data.longOrNull("publishedAt"),
        rejectionReason = data["rejectionReason"]?.let { localizedFrom(it) },
        editorNote = data["editorNote"]?.let { localizedFrom(it) },
        views = data.int("views"),
        reportCount = data.int("reportCount")
    )
}

/* ================================================================= category */

fun Category.toMap(): Map<String, Any?> = mapOf(
    "name" to name.toMap(),
    "emoji" to emoji,
    "isEnabled" to isEnabled
)

fun categoryFrom(id: String, data: Map<String, Any?>): Category = Category(
    id = id,
    name = localizedFrom(data["name"]),
    emoji = data.str("emoji", "📰"),
    // Missing means enabled: a category added by hand without the flag should
    // appear, not silently vanish from the app.
    isEnabled = data.bool("isEnabled", default = true)
)

/* ============================================================= notification */

fun NotificationItem.toMap(): Map<String, Any?> = mapOf(
    "title" to title.toMap(),
    "message" to message.toMap(),
    "timeMillis" to timeMillis,
    "type" to type.name,
    "isRead" to isRead,
    "articleId" to articleId,
    // Stored as the enum name or null. Null means "everyone", which is what the
    // Firestore rule for this collection keys its public read on.
    "targetRole" to targetRole?.name
)

fun notificationFrom(id: String, data: Map<String, Any?>): NotificationItem = NotificationItem(
    id = id,
    title = localizedFrom(data["title"]),
    message = localizedFrom(data["message"]),
    timeMillis = data.long("timeMillis"),
    type = data.enum("type", NotificationType.GENERAL),
    isRead = data.bool("isRead"),
    articleId = data.strOrNull("articleId"),
    targetRole = (data["targetRole"] as? String)?.let { raw ->
        UserRole.entries.firstOrNull { it.name.equals(raw, ignoreCase = true) }
    }
)

/* ======================================================= atomic field updates */

/**
 * An atomic increment, for counters several clients touch at once.
 *
 * View counts and reader reports are written by every reader independently. A
 * read-modify-write would lose counts under concurrency; [FieldValue.increment]
 * is applied server-side.
 */
fun increment(field: String, by: Long = 1): Map<String, Any?> =
    mapOf(field to FieldValue.increment(by))

File: android/src/data/news/newsCodec.ts
/**
 * Ported from core/data/{Firestore,NewsCodec}.kt — exact Firestore field keys
 * for the News module's three collections. `to*()` produces the map written
 * to Firestore; `*From()` decodes a snapshot back into our domain types,
 * degrading to a safe default rather than throwing on a missing/malformed
 * field (mirrors the Kotlin `Map.str/bool/int/long/strList/enum` helpers).
 */
import { DocumentData, DocumentSnapshot, QueryDocumentSnapshot, FieldValue } from '@react-native-firebase/firestore';
import {
  Category,
  NewsArticle,
  NewsStatus,
  NotificationItem,
  NotificationType,
  UserRole,
} from '../../types/news';
import { LocalizedText, toLocalizedText, toLocalizedTextList } from '../../core/i18n/LocalizedText';

export const COLLECTIONS = {
  ARTICLES: 'articles',
  CATEGORIES: 'categories',
  NEWS_NOTIFICATIONS: 'newsNotifications',
  ARTICLE_REPORTS: 'articleReports',
} as const;

type DocSnap = DocumentSnapshot<DocumentData> | QueryDocumentSnapshot<DocumentData>;

function str(data: DocumentData, key: string, fallback = ''): string {
  const v = data[key];
  return typeof v === 'string' ? v : fallback;
}
function strOrNull(data: DocumentData, key: string): string | null {
  const v = data[key];
  return typeof v === 'string' ? v : null;
}
function bool(data: DocumentData, key: string, fallback = false): boolean {
  const v = data[key];
  return typeof v === 'boolean' ? v : fallback;
}
function num(data: DocumentData, key: string, fallback = 0): number {
  const v = data[key];
  return typeof v === 'number' ? v : fallback;
}
function numOrNull(data: DocumentData, key: string): number | null {
  const v = data[key];
  return typeof v === 'number' ? v : null;
}
function strList(data: DocumentData, key: string): string[] {
  const v = data[key];
  return Array.isArray(v) ? v.filter((x): x is string => typeof x === 'string') : [];
}
function localizedOrNull(data: DocumentData, key: string): LocalizedText | null {
  const v = data[key];
  return v == null ? null : toLocalizedText(v);
}

// ---------- articles ----------

export function articleToMap(a: NewsArticle): DocumentData {
  return {
    headline: a.headline,
    shortDescription: a.shortDescription,
    content: a.content,
    categoryId: a.categoryId,
    location: a.location,
    imageUrl: a.imageUrl,
    photoUrls: a.photoUrls,
    videoUrls: a.videoUrls,
    tags: a.tags,
    isBreaking: a.isBreaking,
    isFeatured: a.isFeatured,
    isTrending: a.isTrending,
    status: a.status,
    reporterId: a.reporterId,
    reporterName: a.reporterName,
    reporterAvatarUrl: a.reporterAvatarUrl,
    createdAt: a.createdAt,
    updatedAt: a.updatedAt,
    publishedAt: a.publishedAt,
    rejectionReason: a.rejectionReason,
    editorNote: a.editorNote,
    views: a.views,
    reportCount: a.reportCount,
    // Denormalized, write-only: used for reader ordering without a composite index.
    sortAt: a.publishedAt ?? a.createdAt ?? Date.now(),
  };
}

export function articleFrom(doc: DocSnap): NewsArticle | null {
  const data = doc.data();
  if (!data) return null;
  return {
    id: doc.id,
    headline: toLocalizedText(data.headline),
    shortDescription: toLocalizedText(data.shortDescription),
    content: toLocalizedText(data.content),
    categoryId: str(data, 'categoryId'),
    location: str(data, 'location'),
    imageUrl: str(data, 'imageUrl'),
    photoUrls: strList(data, 'photoUrls'),
    videoUrls: strList(data, 'videoUrls'),
    tags: toLocalizedTextList(data.tags),
    isBreaking: bool(data, 'isBreaking'),
    isFeatured: bool(data, 'isFeatured'),
    isTrending: bool(data, 'isTrending'),
    status: (str(data, 'status', 'DRAFT') as NewsStatus),
    reporterId: str(data, 'reporterId'),
    reporterName: str(data, 'reporterName'),
    reporterAvatarUrl: str(data, 'reporterAvatarUrl'),
    createdAt: num(data, 'createdAt', Date.now()),
    updatedAt: num(data, 'updatedAt', Date.now()),
    publishedAt: numOrNull(data, 'publishedAt'),
    rejectionReason: localizedOrNull(data, 'rejectionReason'),
    editorNote: localizedOrNull(data, 'editorNote'),
    views: num(data, 'views'),
    reportCount: num(data, 'reportCount'),
  };
}

// ---------- categories ----------

export function categoryToMap(c: Category): DocumentData {
  return { name: c.name, emoji: c.emoji, isEnabled: c.isEnabled };
}

export function categoryFrom(doc: DocSnap): Category | null {
  const data = doc.data();
  if (!data) return null;
  return {
    id: doc.id,
    name: toLocalizedText(data.name),
    emoji: str(data, 'emoji', '📰'),
    isEnabled: data.isEnabled == null ? true : bool(data, 'isEnabled', true),
  };
}

// ---------- newsNotifications ----------

export function notificationToMap(n: NotificationItem): DocumentData {
  return {
    title: n.title,
    message: n.message,
    timeMillis: n.timeMillis,
    type: n.type,
    isRead: n.isRead,
    articleId: n.articleId,
    targetRole: n.targetRole,
  };
}

export function notificationFrom(doc: DocSnap): NotificationItem | null {
  const data = doc.data();
  if (!data) return null;
  return {
    id: doc.id,
    title: toLocalizedText(data.title),
    message: toLocalizedText(data.message),
    timeMillis: num(data, 'timeMillis', Date.now()),
    type: (str(data, 'type', 'GENERAL') as NotificationType),
    isRead: bool(data, 'isRead'),
    articleId: strOrNull(data, 'articleId'),
    targetRole: (strOrNull(data, 'targetRole') as UserRole | null),
  };
}

export function increment(by = 1) {
  return FieldValue.increment(by);
}
